### PR TITLE
Allow casts to string for C pointers

### PR DESCRIPTION
**Pull request by nspark**
Allow casts to `string` for both `c_void_ptr` and `c_ptr` objects.

### CODE REVIEW

**Comment by ben-albrecht** [comment]
Hi @nspark -- as always, thanks for the contribution.

Note that Chapel already has an internal `chpldev_refToString()` function [available](https://github.com/chapel-lang/chapel/blob/master/modules/internal/String.chpl#L1778), which will return a string of the pointer integer value, giving similar debugging capabilities. Admittedly, it's a bit ugly. I'd be in favor of modifying this PR to do the following for `c_void_ptr` and `c_ptr`:

* Allow casts to `string`
    * Using `chpldev_refToString()` under the hood
* Make them easily printable
    * By printing the value returned from `chpldev_refToString()` under the hood

This is all assuming that there is no additional value in casting `c_ptr` / `c_void_ptr` to a `uint(64)`. Are there any `uint(64)` operations you had in mind that might be useful beyond printing for debugging purposes?

**Comment by nspark** [comment]
Hi @ben-albrecht, I tried out `chpldev_refToString()` in the following simple example (including my PR branch's changes):

```chapel
var mem = c_calloc(uint(8), 4096);
writeln(chpldev_refToString(mem));                  // prints 0x7819b0
writeln(mem);                                       // prints 0x00007feba6835000
writeln(__primitive("ref to string", mem):string);  // prints 0x7feba6835000
c_free(mem);
```

I don't think I can use `chpldev_refToString()` directly because this prints the address of `mem`, not the address pointed to by `mem`. I (basically) get the behavior I want by calling `__primitive("ref to string")`, although this doesn't seem all the different from casting a pointer to a `uint(64)` via `__primitive("cast")`.

I did happen to notice [this comment](https://github.com/chapel-lang/chapel/blob/master/runtime/src/chplio.c#L29) on `chpl_refToString`, which just made me wonder whether I should be using this primitive at all (or whether it will transparently move to `chpl-string.c`).

Trying to synthesize your recommendation with my understanding of `chpldev_refToString()`, I should:

* add a cast from `c_void_ptr` and `c_ptr` to string
* perform this cast by `__primitive("ref to string")`

Does that seem right? If so, the "preferred" test would then be something like:

```chapel
var mem = c_calloc(uint(8), 4096);
writeln(mem:string);  // prints 0x7feba6835000
c_free(mem);
```

**Comment by ben-albrecht** [comment]
Ah, I was slightly off. Your proposed changes and test case seem correct to me.

File: test/types/cptr/ptr_cast_to_string.chpl [hunk 1]
```diff
@@ -0,0 +1,9 @@
+use Regexp;
+
+var mem = c_calloc(uint(8), 4096);
+if mem != c_nil then
+  assert((mem:string).match(compile("0x[0-9a-f]+")).matched);
+var ptr = mem:c_void_ptr;
+assert((mem:string) == (ptr:string));
+assert((c_nil:string) == "(nil)");
```
Comment (ben-albrecht) [inline test/types/cptr/ptr_cast_to_string.chpl]: tested locally, `c_nil:string` yields `0x0`, instead of `(nil)`.

**Comment by ben-albrecht** [comment]
Should be good to go after fixing the test assertion I pointed out. Did you intend to re-implement `writeThis` methods on the pointer objects? I think they could be useful still.

File: test/types/cptr/ptr_cast_to_string.chpl (hunk 1 above)
Comment (nspark) [inline test/types/cptr/ptr_cast_to_string.chpl]: Hmm. For me on Linux (Ubuntu 16.04 and RHEL6), I see `(nil)`, but I see now that, according to the C specification, the `%p` format produces implementation-defined output. Maybe I should remove that test line.

**Comment by ben-albrecht** [comment]
>  I thought it might be more consistent to require that they be converted to strings to be printable.

Agreed. I'm personally in favor of not requiring the `:string` cast for the sake of productivity (or is it programmability?), but shifting this direction would  call for a broader discussion within the Chapel team.

This discussion shouldn't block the feature proposed here, however. If you fix the test I pointed out, we can merge this now.

**Comment by nspark** [comment]
@ben-albrecht Done!

**Pull request merged**